mvn package -Pjlink-image is failing on Windows

System Info.:    
    
    Helidon MP 4.1.2
    OpenJdk 22 (2024-03-19) - Java version: 22, vendor: Oracle Corporation
    Maven 3.8.1
    OS name: "windows 11", version: "10.0"

It looks like a execution permission issue, wondering if there is an alternative or flag available to overcome the problem?

I am getting the following error:

    [INFO] ------------------------------------------------------------------------
    [INFO] BUILD FAILURE
    [INFO] ------------------------------------------------------------------------
    [INFO] Total time:  13.689 s
    [INFO] Finished at: 2024-10-16T16:31:10-04:00
    [INFO] ------------------------------------------------------------------------
    [ERROR] Failed to execute goal io.helidon.build-tools:helidon-maven-plugin:4.0.12:jlink-image (jlink-image) on project java-modules-helidon: Image generation error: java.nio.file.AccessDeniedException: \java-modules-helidon\target\java-modules-helidon-jri\lib\start.jsa -> [Help 1]
    org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal io.helidon.build-tools:helidon-maven-plugin:4.0.12:jlink-image (jlink-image) on project java-modules-helidon: Image generation error
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
        at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
        at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
        at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
        at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
        at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
        at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
        at java.lang.reflect.Method.invoke (Method.java:580)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    Caused by: org.apache.maven.plugin.MojoExecutionException: Image generation error
        at io.helidon.build.maven.JLinkImageMojo.execute (JLinkImageMojo.java:152)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
        at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
        at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
        at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
        at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
        at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
        at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
        at java.lang.reflect.Method.invoke (Method.java:580)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    Caused by: java.io.UncheckedIOException: java.nio.file.AccessDeniedException: \java-modules-helidon\target\java-modules-helidon-jri\lib\start.jsa
        at io.helidon.build.common.FileUtils.lambda$deleteDirectory$3 (FileUtils.java:431)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept (ForEachOps.java:184)
        at java.util.ArrayList.forEach (ArrayList.java:1597)
        at java.util.stream.SortedOps$RefSortingSink.end (SortedOps.java:395)
        at java.util.stream.Sink$ChainedReference.end (Sink.java:261)
        at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:557)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:546)
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential (ForEachOps.java:151)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential (ForEachOps.java:174)
        at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:265)
        at java.util.stream.ReferencePipeline.forEach (ReferencePipeline.java:611)
        at io.helidon.build.common.FileUtils.deleteDirectory (FileUtils.java:427)
        at io.helidon.build.linker.util.JavaRuntime.prepareJriDirectory (JavaRuntime.java:119)
        at io.helidon.build.linker.Configuration$Builder.build (Configuration.java:521)
        at io.helidon.build.maven.JLinkImageMojo.execute (JLinkImageMojo.java:149)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
        at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
        at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
        at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
        at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
        at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
        at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
        at java.lang.reflect.Method.invoke (Method.java:580)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    Caused by: java.nio.file.AccessDeniedException: \java-modules-helidon\target\java-modules-helidon-jri\lib\start.jsa
        at sun.nio.fs.WindowsException.translateToIOException (WindowsException.java:89)
        at sun.nio.fs.WindowsException.rethrowAsIOException (WindowsException.java:103)
        at sun.nio.fs.WindowsException.rethrowAsIOException (WindowsException.java:108)
        at sun.nio.fs.WindowsFileSystemProvider.implDelete (WindowsFileSystemProvider.java:273)
        at sun.nio.fs.AbstractFileSystemProvider.delete (AbstractFileSystemProvider.java:104)
        at java.nio.file.Files.delete (Files.java:1152)
        at io.helidon.build.common.FileUtils.lambda$deleteDirectory$3 (FileUtils.java:429)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept (ForEachOps.java:184)
        at java.util.ArrayList.forEach (ArrayList.java:1597)
        at java.util.stream.SortedOps$RefSortingSink.end (SortedOps.java:395)
        at java.util.stream.Sink$ChainedReference.end (Sink.java:261)
        at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:557)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:546)
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential (ForEachOps.java:151)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential (ForEachOps.java:174)
        at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:265)
        at java.util.stream.ReferencePipeline.forEach (ReferencePipeline.java:611)
        at io.helidon.build.common.FileUtils.deleteDirectory (FileUtils.java:427)
        at io.helidon.build.linker.util.JavaRuntime.prepareJriDirectory (JavaRuntime.java:119)
        at io.helidon.build.linker.Configuration$Builder.build (Configuration.java:521)
        at io.helidon.build.maven.JLinkImageMojo.execute (JLinkImageMojo.java:149)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
        at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
        at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
        at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
        at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
        at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
        at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
        at java.lang.reflect.Method.invoke (Method.java:580)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    [ERROR]
    [ERROR]
    [ERROR] For more information about the errors and possible solutions, please read the following articles:
    [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

--- a/linker/src/main/java/io/helidon/build/linker/ClassDataSharing.java
+++ b/linker/src/main/java/io/helidon/build/linker/ClassDataSharing.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2019, 2022 Oracle and/or its affiliates.
+ * Copyright (c) 2019, 2024 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -365,7 +365,7 @@ public final class ClassDataSharing {
             }
             if (Constants.OS == OSType.Windows) {
                 // Try to make the archive file writable so that a second run can delete the image
-                archiveFile.toFile().setWritable(true);
+                jri.resolve(archiveFile).toFile().setWritable(true);
             }
         }
 

